front: round trips in nge to osrd

Synar opened the pull request:
Close [#osrd-confidential/1022](https://github.com/osrd-project/osrd-confidential/issues/1022)

louisgreiner commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -449,22 +451,56 @@ const handleCreateTimetableItem = async (
-  ).unwrap();
-  if (newTimetableItems.length === 0) {
-    throw new Error('Failed to create paced train');
+  if (trainrun.trainrunDirection === 'one_way') {
+    const newTimetableItems = await dispatch(
+      osrdEditoastApi.endpoints.postTimetableByIdPacedTrains.initiate({
+        id: timetableId,
+        body: [pacedTrain],
+      })
+    ).unwrap();
+
+    if (newTimetableItems.length !== 1) {
+      throw new Error('Failed to create paced train');
+    }
+    const newPacedTrain: PacedTrainWithPacedTrainId = {
+      ...newTimetableItems[0],
+      id: formatEditoastIdToPacedTrainId(newTimetableItems[0].id),
+    };
+    state.timetableItemIdByNgeId.set(trainrun.id, [newPacedTrain.id, null]);
+    addUpsertedTimetableItems([newPacedTrain]);
```
This case is likely never gonna happen, since the default `Direction` value of a newly created `Trainrun` in NGE is `round_trip`. However, we never know if this default value could change one day, so it's nice to have it. Maybe a small comment for that would be nice 👍

Synar replied:
Done ^^

emersion commented on front/src/utils/trainId.ts:
```diff
@@ -144,6 +145,17 @@ export const extractEditoastIdFromPacedTrainId = (pacedTrainId: PacedTrainId): n
 };
 
+/**
+ * Given a timetable item id with a TimetableItemId format (used across the front),
+ * returns the train id in the Editoast format (used for api).
```
This function is dangerous: it returns an ID which is either a paced train ID, either a train schedule ID, without giving more details. These two must not be mixed up: train schedule ID 42 and paced train ID 42 can coexist and they refer to completely separate objects.

I believe we should drop this function and stick to the existing ones.

Synar replied:
It was quite practical though, for example to shorten deleteTimetableItemById code a bit. But sure, I suppose I can inline it.

Synar replied:
It does makes me think I should double check the behavior when converting from paced to train schedule when a return is present.

emersion replied:
I don't like storing multiple kinds of objects in the same variable because mixups become possible. For instance, I prefer:

```ts
if (isPacedTrainId(timetableItemId)) {
  const id = extractEditoastIdFromPacedTrainId(timetableItemId);
  await deletePacedTrain(id);
} else {
  const id = extractEditoastIdFromTrainScheduleId(timetableItemId);
  await deleteTrainSchedule(id);
}
```

over:

```ts
const id = isPacedTrainId(timetableItemId)
  ? extractEditoastIdFromPacedTrainId(timetableItemId)
  : extractEditoastIdFromTrainScheduleId(timetableItemId);
await isPacedTrainId(timetableItemId)
  ? deletePacedTrain(id);
  : deleteTrainSchedule(id);
```

The second version conditionally stores one kind of object or the other into a single variable. It's a bit like storing a number of people and a speed in m/s in the same variable based on a condition. If this piece of code evolves, e.g. these two conditions become farther apart, TypeScript will not warn if the two conditions become out-of-sync. The `TimetableItemId` type exists exactly to not mix up different kinds of IDs.

Ideally, we'd convert from a `TimetableItemId` to a raw editoast ID inside rtk-query, like we do for `getTimetableItemById` for instance, or have helpers such as:

https://github.com/OpenRailAssociation/osrd/blob/8c312e8203fe97863b157f4e4337447aebe336aa/front/src/modules/timetableItem/helpers/updateTimetableItemHelpers.ts#L99

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -449,22 +451,59 @@ const handleCreateTimetableItem = async (
-  if (newTimetableItems.length === 0) {
-    throw new Error('Failed to create paced train');
+  if (trainrun.direction === 'one_way') {
+    // This case should no longer happen in practice, as by default nge should create trains as round trips
```
Instead of keeping dead code, can we throw an error if this happens to be a one-way trainrun?

Synar replied:
Eh, I don't like being too dependent on something that is nge config, and that could change in the future. But I suppose it would not be too hard to fix in that case.

emersion replied:
We already depend on the fact that NGE creates a paced train by default.

Synar replied:
Do we? I believe there was some back and forth when I was developing about whether we should have the fork overrule nge to create a trains schedule instead, and I think that got implemented in one version.

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -449,22 +451,59 @@ const handleCreateTimetableItem = async (
+      osrdEditoastApi.endpoints.postRoundTripsPacedTrains.initiate({
+        roundTrips: { round_trips: [[newTimetableItems[0].id, newTimetableItems[1].id]] },
+      })
+    );
```
We should `unwrap()` here to ensure errors aren't silently ignored (ie, throw an error instead of doing nothing).

Synar replied:
Sure!

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -506,6 +504,57 @@ const handleCreateTimetableItem = async (
+      await deleteTimetableItemById(timetableItemId, dispatch, addDeletedTimetableItemIds);
+    }
+  }
+  if (timetableItemIds?.[1])
```
Style: can we add brackets? I find the bracket-less style to be fine for one-liners, but the code below is longer.

Synar replied:
I can't find the code this refers to, was this part of the deleted delete code?

emersion replied:
Maybe "front: ngetoosrd: handle round trips on timetable item update"?

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -506,6 +504,57 @@ const handleCreateTimetableItem = async (
+  addDeletedTimetableItemIds([timetableItemId]);
+};
+
+const handleDeleteTimetableItem = async (
```
In this commit, `handleDeleteTimetableItem` still exists below.

Synar replied:
Which commit are you talking about?

Synar replied:
Oh the commit that moved the function got split in two, that is painful.

emersion replied:
It sounds like this comment is still relevant. In "front: ngetoosrd: handle round trips on timetable item delete", two versions of `handleDeleteTimetableItem` are defined.

This commit should modify the existing function instead of creating a new one.

emersion replied:
I've made the change in this branch: [emr/ali/nge-to-osrd-round-trips](https://github.com/OpenRailAssociation/osrd/commits/emr/ali/nge-to-osrd-round-trips/)

While at it, I've fixed the `Co-authored-by` lines in commit messages (which had a typo: `Co-Authored` is not recognized by GitHub) and ensured all commits build properly.

If it's to your liking, feel free to adopt it (`git reset --hard emr/ali/nge-to-osrd-round-trips`).

louisgreiner replied:
amazing, thanks (and sorry for the typo!)

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -506,6 +504,57 @@ const handleCreateTimetableItem = async (
+      await deleteTimetableItemById(timetableItemId, dispatch, addDeletedTimetableItemIds);
+    }
+  }
+  if (timetableItemIds?.[1])
```
No need to delete the roundtrips if we're deleting the trains. Deleting the trains automatically deletes any associated roundtrips.

Synar replied:
I coded this before the back was implemented, there was a fixup commit dropping it already. I did not rebase it just in case =p

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -562,13 +563,12 @@ const handleUpdateTimetableItem = async ({
-  // TODO: handle roundtrips
-  const timetableItemId = state.timetableItemIdByNgeId.get(trainrun.id)![0];
-  const timetableItem = await fetchTimetableItem(timetableItemId, dispatch);
+  const timetableIds = state.timetableItemIdByNgeId.get(trainrun.id)!;
```
This name makes it sounds like this is a timetable ID, when it's a timetable item ID. Could we keep the old name?

Synar replied:
Sure thanks! It was a typo when going from one id to an id table.

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -602,63 +602,179 @@ const handleUpdateTimetableItem = async ({
-  addDeletedTimetableItemIds([timetableItemId]);
-  state.timetableItemIdByNgeId.delete(trainrunId);
+  if (trainrun.direction === 'one_way') {
+    if (timetableIds[1]) {
+      // NGE always selects the forward trip by default when going from round trip to one way trip,
+      // thus the trip that needs to be deleted is always the return trip
+      if (isPacedTrainId(timetableItemToUpdate.id))
+        await dispatch(
+          osrdEditoastApi.endpoints.postRoundTripsPacedTrains.initiate({
+            roundTrips: {
+              one_ways: [extractEditoastIdFromTimetableItemId(timetableItemToUpdate.id)],
+            },
+          })
+        );
+      else
+        await dispatch(
+          osrdEditoastApi.endpoints.postRoundTripsTrainSchedules.initiate({
+            roundTrips: {
+              one_ways: [extractEditoastIdFromTimetableItemId(timetableItemToUpdate.id)],
+            },
+          })
+        );
```
I think this can be dropped since we delete the return trip below.

Synar replied:
I don't believe so. When testing, I believe editoast drops the whole line instead of converting to one way when dropping a return.

emersion replied:
Oh, you're right!

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -602,63 +602,179 @@ const handleUpdateTimetableItem = async ({
+
+  if (timetableIds[1]) {
+    // update return if already present
+    if (paced) {
```
Nit: we could change this check to `if (updatedPacedTrain)` so that we don't need to use the `!` operator below.

Synar replied:
Sure!

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -602,63 +602,179 @@ const handleUpdateTimetableItem = async ({
+      );
+    }
+  } else {
+    /// otherwise create return
```
Style nit: an extra `/` made its way here

Synar replied:
Thanks!

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -602,63 +602,179 @@ const handleUpdateTimetableItem = async ({
+      };
+
+      const newTimetableItems = await dispatch(
+        osrdEditoastApi.endpoints.postTimetableByIdPacedTrains.initiate({
```
We already have a helper named `createPacedTrain()`, would it make sense to re-use it here?

Synar replied:
I was not aware these helpers existed, they would probably fit the code very well. Thanks!

Synar replied:
Hmm, now I kind of want to generalize it to take a list of trains and use it in the handleCreate too. Not sure.

emersion commented on front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -602,63 +602,179 @@ const handleUpdateTimetableItem = async ({
+            ],
+          },
+        })
+      );
```
Ditto: we're missing `.unwrap()` here.

emersion reviewed (commented):
Thanks a lot for splitting the commits!

Synar commented:
@emersion Should be done. I also simplified some things a bit and tried a new naming convention for the nge/osrd timetableitem to/for update.

Also I think the graph walk was bugged, should be fixed now. The tests now cover one one way train and one round trip train.

louisgreiner commented:
Had to rewrite the history in order to rebase on `dev` :s I hope I didn't mess up

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -546,36 +549,35 @@ const handleUpdateTimetableItem = async ({
-  const timetableItemId = state.timetableItemIdByNgeId.get(trainrun.id)![0];
-  const timetableItem = await fetchTimetableItem(timetableItemId, dispatch);
+  const timetableItemIds = state.timetableItemIdByNgeId.get(trainrun.id)!;
+  const osrdForwardTimetableItem = await fetchTimetableItem(timetableItemIds[0], dispatch);
```
Nit: I don't think it's necessary to put "osrd" in the name, since "timetableItem" is an OSRD concept in the first place (as opposed to "trainrun").

Ah, I now see an "nge" variant below. If this is named this way to convey the fact that this is the old timetable item, and we're creating the new/updated one below, maybe "old" and "new" would be more fitting instead of "osrd"/"nge"?

louisgreiner replied:
agreed! done

Synar replied:
The idea was that we had itemToUpdate, itemForUpdate, and updatedItem, which was quite unclear. So they became osrdItem, ngeItem, and the final updated one. But I've now realized that I ended up dropping entirely the updatedItem, so now simply old and new should be fine indeed.

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd.ts:
```diff
@@ -584,64 +586,174 @@ const handleUpdateTimetableItem = async ({
+    if (timetableItemIds[1]) {
+      // NGE always selects the forward trip by default when going from round trip to one way trip,
+      // thus the trip that needs to be deleted is always the return trip
+      if (isPacedTrainId(osrdForwardTimetableItem.id))
```
Nit: since the condition body uses multiple lines, can we add brackets?

louisgreiner replied:
Done

emersion reviewed (commented):
Just a few minor comments left!

Commit:
front: ngetoosrd: handle round trips on timetable item creation

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Commit:
front: ngetoosrd: handle round trips on timetable item deletion

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Commit:
front: ngetoosrd: handle round trips on timetable item update

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Commit:
front: ngetoosrd: always consider nge stops as stops

When lacking a departure time, nge stops (node transitions marked with isNoneStopTransit set to false)
were not considered as stops by osrd, and the stop_for property of the corresponding
path steps was set to null.

This could lead to a return trip not having the same stops as the forward trip,
leading to an incompatibility.

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>
Co-authored-by: Simon Ser <contact@emersion.fr>

Commit:
front: osrdtonge: set return departure time same as forward

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>
Co-authored-by: Simon Ser <contact@emersion.fr>

Commit:
front: ngetoosrd: fix and simplify sections ordering

getTrainrunSectionsByTrainrunId was walking through the sections graph,
trying to impose both a consistent order as well as forcing trips to go
from source to target.

This worked well when we only had one way trains in osrd, and round trips in nge.

However, nge now naturally handles one way trips, as well as now enforces
source target ordering itself, meaning we no longer need to handle it.

Additionally, when changing the direction of one way trip in nge,
it could get reversed twice, once by nge and once by getTrainrunSectionsByTrainrunId.

Thus always assume that source to target ordering is already correct in getTrainrunSectionsByTrainrunId
and stop modifying it.

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>
Co-authored-by: Simon Ser <contact@emersion.fr>

Commit:
front: ngetoosrd: handle one way trips in convertNgeDtoToOsrd + bump NGE package

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Co-authored-by: Alice Khoudli <alice.khoudli@polytechnique.org>

louisgreiner reviewed (commented):
Also rewrote the commit history to try to update the existing `handleDeleteTimetableItem`, but I could not do it since it is now used before update logic is defined.
@emersion

SarahBellaha reviewed (approved):
LGTM!

Commit:
front: call refreshNge when round trips are added/removed

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>

emersion reviewed (approved):
\o/

Synar reviewed (commented):
Nice thanks! 💜
